File: PA46.java

// Leetcode 113. Path Sum II
// Time complexity: O(n) // As we are traversing each node only once
//Space Complexity : O(n) // n is the no of nodes in tree. Space is used for stack.
class Solution {
    List<List<Integer>> res;
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        res= new ArrayList<List<Integer>>();
        
        if(root==null){
            return res;
        }
        
        dfs(root,sum,0, new ArrayList<Integer>());
        return res;
    }
    
   public void dfs(TreeNode root, int sum,int tsum, List<Integer> tres){
        if(root==null){
            return ;
        }
       
        if(root.left==null && root.right==null && tsum+root.val ==sum){
                tres.add(root.val);
                List<Integer> l= new ArrayList<>(tres);
                res.add(l);
                 tres.remove(tres.size()-1);
        }
        tres.add(root.val);
        tsum=tsum+root.val;
        dfs(root.left,sum,tsum,tres);   
        dfs(root.right,sum,tsum,tres);
        tres.remove(tres.size()-1);
       
    }
}

File: PA_46.py
#LC - 113. Path Sum II
#Time Complexity - O(n)
#Space Complexity - O(n)
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def pathSum(self, root, sun):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        if not root:
            return []
        self.ans = []
        def helper(root,path,tsum,sun):
            if root:
                tsum = tsum + root.val
                path.append(root.val)
                if not root.left and not root.right:
                    if tsum == sun:
                        x = [i for i in path]
                        self.ans.append(x)
                helper(root.left,path,tsum,sun)
                helper(root.right,path,tsum,sun)
                path.pop()
        helper(root,[],0,sun)
        return(self.ans)